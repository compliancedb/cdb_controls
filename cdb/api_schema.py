class ApiSchema:
    @staticmethod
    def url_for_owner_projects(host, project_data):
        return host + '/api/v1/projects/' + project_data["owner"] + "/"

    @staticmethod
    def url_for_project(host, project_data):
        return ApiSchema.url_for_owner_projects(host, project_data) + project_data["name"]

    @staticmethod
    def url_for_releases(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/releases/'

    @staticmethod
    def url_for_commit(host, project_data, commit):
        return ApiSchema.url_for_project(host, project_data) + '/commits/' + commit

    @staticmethod
    def url_for_artifacts(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/artifacts/'

    @staticmethod
    def url_for_artifact(host, project_data, sha256_digest):
        return ApiSchema.url_for_artifacts(host, project_data) + sha256_digest

    @staticmethod
    def url_for_artifacts(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/artifacts/'

    @staticmethod
    def url_for_release(host, project_data, release_number):
        return ApiSchema.url_for_releases(host, project_data) + release_number