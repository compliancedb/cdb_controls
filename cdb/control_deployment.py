def control_deployment_approved(approvals):
    for approval in approvals:
        if approval["state"] == "APPROVED":
            return True
    return False
