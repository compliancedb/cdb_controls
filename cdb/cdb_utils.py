import getopt
import json
import os
import sys
import docker
import requests as req
from junitparser import JUnitXml
from requests.auth import HTTPBasicAuth

from cdb.settings import CDB_SERVER

CMD_HELP = 'ensure_project.py -p <project.json>'


def project_exists_in_cdb(project_data, projects):
    exists = False
    for project in projects:
        if project["name"] == project_data["name"]:
            exists = True
    return exists


def get_project_list_from_cdb(projects_url):
    resp = req.get(projects_url)
    # If the response was successful, no Exception will be raised
    resp.raise_for_status()
    print(resp.text)
    projects = resp.json()
    return projects


def get_project_from_cdb(project_url):
    resp = req.get(project_url)
    # If the response was successful, no Exception will be raised
    resp.raise_for_status()
    print(resp.text)
    project = resp.json()
    return project


def load_project_configuration(json_data_file):
    project_data = json.load(json_data_file)
    return project_data


def create_artifact(api_token, host, project_config_file, sha256, filename, description, git_commit, commit_url, build_url,
                    is_compliant):
    project_data = load_project_configuration(project_config_file)

    '''
    curl -H 'Content-Type: application/json' \
     -X PUT \
     -d '{"sha256": "'"$3"'", "filename": "'"$4"'", "description": "'"$5"'", "git_commit": "'"$6"'", "commit_url": "'"$7"'", "build_url": "'"$8"'", "is_compliant": "true"}' \
    http://server:8001/api/v1/projects/$1/$2/artifacts/
    '''

    create_artifact_payload = {
        "sha256": sha256,
        "filename": filename,
        "description": description,
        "git_commit": git_commit,
        "commit_url": commit_url,
        "build_url": build_url,
        "is_compliant": is_compliant
    }
    url = url_for_artifacts(host, project_data)
    put_payload(create_artifact_payload, url, api_token)


def put_payload(payload, url, api_token):
    headers = {"Content-Type": "application/json"}
    print("Putting this payload:")
    print(json.dumps(payload, sort_keys=True, indent=4))
    print("To url: " + url)
    resp = req.put(url, data=json.dumps(payload), headers=headers, auth=HTTPBasicAuth(api_token, 'unused'))
    print(resp.text)


def url_for_artifacts(host, project_data):
    return url_for_project(host, project_data) + project_data["name"] + '/artifacts/'


def add_evidence(api_token, host, project_file_contents, sha256_digest, evidence):
    project_data = load_project_configuration(project_file_contents)
    url_for_artifact = url_for_artifacts(host, project_data) + sha256_digest

    put_payload(evidence, url_for_artifact, api_token)


def url_for_project(host, project_data):
    return host + '/api/v1/projects/' + project_data["owner"] + '/'


def rchop(thestring, ending):
    if thestring.endswith(ending):
        return thestring[:-len(ending)]
    return thestring


def parse_cmd_line():
    project_file = ""
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hp:", ["project="])
    except getopt.GetoptError:
        print(CMD_HELP)
        sys.exit(2)
    for opt, arg in opts:
        print(opt + ": " + arg)
        if opt == '-h':
            print(CMD_HELP)
            sys.exit()
        elif opt in ("-p", "--project"):
            project_file = arg
    return project_file


def get_image_details():
    docker_image = os.getenv('CDB_DOCKER_IMAGE', "NO_DOCKER_IMAGE_FOUND")
    sha256_digest = os.getenv('CDB_ARTIFACT_SHA', None)
    if sha256_digest is None:
        client = docker.from_env()
        print("Inspecting docker image for sha256Digest")
        image = client.images.get(docker_image)
        sha256_digest = image.attrs["RepoDigests"][0].split(":")[1]
    else:
        print("Docker image digest found from environment variable")

    print("IMAGE DIGEST: " + str(image.attrs["RepoDigests"][0]))
    return docker_image, sha256_digest


def env_is_compliant():
    return os.getenv('CDB_IS_COMPLIANT', "FALSE") == "TRUE"


def is_compliant_suite(junit_xml):
    if junit_xml.failures != 0:
        return False, "Tests contain failures"
    if junit_xml.errors != 0:
        return False, "Tests contain errors"
    return True, "All tests passed"


def load_test_results(file_path):
    test_xml = JUnitXml.fromfile(file_path)
    return test_xml


def is_compliant_test_results(file_path):
    """
    This parses a junit xml file to determine if there are any errors or failures

    return:
        A tuple, with is_compliant, plus a message string
    """
    test_xml = load_test_results(file_path)
    if test_xml._tag == "testsuites":
        for suite in test_xml:
            suite_is_compliant, message = is_compliant_suite(suite)
            if not suite_is_compliant:
                return suite_is_compliant, message
        # every test suite passed, so return True
        return True, "All tests passed"
    if test_xml._tag == "testsuite":
        return is_compliant_suite(test_xml)
    return False, "Could not find test suite(s)"


def control_junit():
    print("Publish evidence to ComplianceDB")
    junit_results_path = "/data/junit/junit.xml"

    is_compliant, message = is_compliant_test_results(junit_results_path)
    evidence_type = os.getenv('CDB_EVIDENCE_TYPE', "junit")
    description = "JUnit results xml verified by compliancedb/cdb_controls: " + message
    build_url = os.getenv('CDB_CI_BUILD_URL', "URL_UNDEFINED")

    evidence = build_evidence_dict(is_compliant, evidence_type, description, build_url)
    send_evidence(evidence)


def put_evidence():
    print("Publish evidence to ComplianceDB")

    is_compliant = env_is_compliant()
    evidence_type = os.getenv('CDB_EVIDENCE_TYPE', "EVIDENCE_TYPE_UNDEFINED")
    description = os.getenv('CDB_DESCRIPTION', "UNDEFINED")
    build_url = os.getenv('CDB_CI_BUILD_URL', "URL_UNDEFINED")

    evidence = build_evidence_dict(is_compliant, evidence_type, description, build_url)
    send_evidence(evidence)


def send_evidence(evidence):
    print(evidence)
    project_file = parse_cmd_line()
    with open(project_file) as project_file_contents:
        _docker_image_unused, sha256_digest = get_image_details()
        api_token = os.getenv('CDB_API_TOKEN', 'NO_API_TOKEN_DEFINED')
        host = os.getenv('CDB_HOST', CDB_SERVER)
        add_evidence(api_token, host, project_file_contents, sha256_digest, evidence)


def build_evidence_dict(is_compliant, evidence_type, description, build_url):
    evidence = {"evidence_type": evidence_type, "contents": {
        "is_compliant": is_compliant,
        "url": "",
        "description": ""
    }}
    evidence["contents"]["description"] = description
    evidence["contents"]["url"] = build_url
    return evidence
