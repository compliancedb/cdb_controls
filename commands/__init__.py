from .command_error import CommandError

from .json_loader import load_json

from .context import Context

from .command import Command

from .declare_pipeline import DeclarePipeline
from .control_deployment import ControlDeployment
from .log_artifact import LogArtifact
from .log_deployment import LogDeployment
from .log_evidence import LogEvidence
from .log_approval import LogApproval
from .log_test import LogTest

from .command_builder import build_command, COMMANDS
from .runner import run, main
