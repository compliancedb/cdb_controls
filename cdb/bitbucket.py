import json
import os
import sys

import requests

from cdb.cdb_utils import build_evidence_dict, send_evidence


def put_bitbucket_pull_request(project_config_file):

    pull_requests = get_pull_request_for_current_commit()

    print("Found pull request info: " + str(pull_requests))
    evidence = build_evidence_dict(
        is_compliant=True,
        evidence_type="pull_request",
        description="Bitbucket pull request",
        build_url=pull_requests[0]["pullRequestURL"])
    evidence["contents"]["source"] = pull_requests
    send_evidence(project_config_file, evidence)


def get_pull_request_for_current_commit():
    workspace = os.getenv('BITBUCKET_WORKSPACE', None)
    repository = os.getenv('BITBUCKET_REPO_SLUG', None)
    commit = os.getenv('BITBUCKET_COMMIT', None)
    user = os.getenv('BITBUCKET_USER', None)
    password = os.getenv('BITBUCKET_PWD', None)
    pull_requests = get_pull_requests_from_bitbucket_api(
        workspace=workspace, repository=repository, commit=commit, username=user, password=password)
    return pull_requests


def get_pull_requests_from_bitbucket_api(workspace, repository, commit, username, password):
    pull_requests_evidence = []

    url = f"https://api.bitbucket.org/2.0/repositories/{workspace}/{repository}/commit/{commit}/pullrequests"
    r = requests.get(url, auth=(username, password))
    if r.status_code == 200:
        pull_requests_json = json.loads(r.text)
        pull_requests = pull_requests_json["values"]
        for pr in pull_requests:
            pr_evidence = {}

            pr_html_url = pr['links']['html']['href']
            pr_api_url = pr['links']['self']['href']

            pr_evidence['pullRequestMergeCommit'] = commit
            pr_evidence['pullRequestURL'] = pr_html_url

            pr_evidence = get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password)

            pull_requests_evidence.append(pr_evidence)

    elif r.status_code == 202:
        print("Repository pull requests are still being indexed, please retry.")
        sys.exit(1)
    elif r.status_code == 404:
        print("Repository does not exists or pull requests are not indexed. Please make sure Pull Request Commit Links app is installed")
        sys.exit(2)
    else:
        print(f"Exception occured in fetching pull requests. Http return code is {r.status_code}")
        sys.exit(3)

    return pull_requests_evidence


def get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password):
    r = requests.get(pr_api_url, auth=(username, password))
    if r.status_code == 200:
        pr_json = json.loads(r.text)
        pr_evidence['pullRequestState'] = pr_json['state']
        participants = pr_json['participants']
        approvers = ""
        if len(participants) > 0:
            for participant in participants:
                if participant['approved']:
                    approvers = approvers + participant['user']['display_name'] + ","
            approvers = approvers[:-1]
            pr_evidence['approvers'] = approvers
        else:
            print("No approvers found")
    else:
        print("Error occured in fetching pull request details. Please review repository permissions.")
    return pr_evidence





