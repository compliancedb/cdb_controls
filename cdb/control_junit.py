#!/usr/bin/env python

from cdb.cdb_utils import control_junit

if __name__ == '__main__':
    control_junit()
