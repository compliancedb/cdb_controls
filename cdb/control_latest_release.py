#!/usr/bin/env python

from cdb.cdb_utils import control_latest_release

if __name__ == '__main__':
    control_latest_release()
