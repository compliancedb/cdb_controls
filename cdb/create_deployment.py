#!/usr/bin/env python
import os

from cdb.api_schema import ApiSchema
from cdb.cdb_utils import set_artifact_sha_env_variable_from_file_or_image, load_user_data, parse_cmd_line, \
    send_evidence, load_project_configuration, get_host, get_api_token
from cdb.http import http_post_payload


def create_deployment_payload(env=os.environ):
    deployment = {}
    deployment["artifact_sha256"] = env.get('CDB_ARTIFACT_SHA', None)
    deployment["environment"] = env.get('CDB_ENVIRONMENT', None)
    deployment["description"] = env.get('CDB_DESCRIPTION', None)

    if env.get("CDB_USER_DATA"):
        deployment["user_data"] = load_user_data()
    return deployment


def create_deployment(project_file):
    # Todo: make this smoother
    set_artifact_sha_env_variable_from_file_or_image()
    payload = create_deployment_payload()

    with open(project_file) as json_data_file:
        project_data = load_project_configuration(json_data_file)
        url = ApiSchema.url_for_deployments(get_host(), project_data)
        http_post_payload(payload, url, get_api_token())





"""
| CDB_HOST | Optional | The host name for ComplianceDB, default is https://app.compliancedb.com |
| CDB_API_TOKEN | Required | Your API token for ComplianceDB |
| CDB_ARTIFACT_SHA or CDB_ARTIFACT_DOCKER_IMAGE or CDB_ARTIFACT_FILE | Required | The artifact sha that is being deployed |
| CDB_ENVIRONMENT | Required | The environment the artifact is being deployed to |
| CDB_DESCRIPTION | Optional | A description for the deployment |
| CDB_USER_DATA_FILE | Optional | The user data to embed in the deployment, if any (should be mounted in the container) |
"""
if __name__ == '__main__':
    project_file = parse_cmd_line()
    create_deployment(project_file)
