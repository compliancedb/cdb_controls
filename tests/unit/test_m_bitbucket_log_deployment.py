from cdb.create_deployment import create_deployment
from commands import run, External

from tests.utils import *

APPROVAL_DIR = "tests/unit/approved_executions"
APPROVAL_FILE = "test_m_bitbucket_log_deployment"

BB = 'bitbucket.org'
BB_ORG = 'acme'
BB_REPO = 'road-runner'
COMMIT = "abc50c8a53f79974d615df335669b59fb56a4ed3"
BUILD_NUMBER = '1975'

DOMAIN = "app.compliancedb.com"
OWNER = "acme"
PIPELINE = "road-runner"
API_TOKEN = "5199831f4ee3b79e7c5b7e0ebe75d67aa66e79d4"

IMAGE_NAME = "acme/road-runner:4.67"
SHA256 = "aacdaef69c676c2466571d3288880d559ccc2032b258fc5e73f99a103db462ee"

DESCRIPTION = f"Deployment of {IMAGE_NAME}:aacdaef to docker hub"
ENVIRONMENT = "dockerhub"
USER_DATA = "/app/tests/data/user_data.json"
USER_DATA_JSON = {'status': 'deployed'}


def test_bitbucket(capsys, mocker):
    # The original bitbucket code did not do a BitBucket translation for create_deployment
    env = old_log_deployment_env()
    set_env_vars = {'CDB_ARTIFACT_SHA': SHA256}

    with dry_run(env, set_env_vars):
        mocker.patch('cdb.cdb_utils.calculate_sha_digest_for_docker_image', return_value=SHA256)
        create_deployment("tests/data/Merkelypipe.acme-roadrunner.json")

    verify_approval(capsys, ["out"])

    # extract data from approved cdb text file
    import inspect
    this_test = inspect.stack()[0].function
    approved = f"{APPROVAL_DIR}/{APPROVAL_FILE}.{this_test}.approved.txt"
    with open(approved) as file:
        old_approval = file.read()
    _old_blurb, old_method, old_payload, old_url = extract_blurb_method_payload_url(old_approval)

    expected_method = "Posting"
    expected_url = f"https://{DOMAIN}/api/v1/projects/{OWNER}/{PIPELINE}/deployments/"
    expected_payload = {
        "artifact_sha256": SHA256,
        "build_url": f"https://{BB}/{BB_ORG}/{BB_REPO}/addon/pipelines/home#!/results/{BUILD_NUMBER}",
        "description": DESCRIPTION,
        "environment": ENVIRONMENT,
        "user_data": USER_DATA_JSON,
    }

    # verify data from approved cdb text file
    assert old_method == expected_method
    assert old_url == expected_url
    assert old_payload == expected_payload

    # make merkely call
    ev = new_log_deployment_env()
    with dry_run(ev) as env:
        with MockDockerFingerprinter(IMAGE_NAME, SHA256) as fingerprinter:
            external = External(env=env, docker_fingerprinter=fingerprinter)
            method, url, payload = run(external)

    # verify matching data
    assert method == expected_method
    assert url == expected_url
    assert payload == expected_payload

    assert extract_blurb(capsys_read(capsys)) == [
        'MERKELY_COMMAND=log_deployment',
    ]


def old_log_deployment_env():
    return {
        "CDB_HOST": "https://app.compliancedb.com",
        "CDB_API_TOKEN": "5199831f4ee3b79e7c5b7e0ebe75d67aa66e79d4",
        "CDB_ARTIFACT_DOCKER_IMAGE": "acme/runner:4.56",
        "CDB_ENVIRONMENT": ENVIRONMENT,
        "CDB_DESCRIPTION": DESCRIPTION,
        "CDB_CI_BUILD_URL": f"https://{BB}/{BB_ORG}/{BB_REPO}/addon/pipelines/home#!/results/{BUILD_NUMBER}",
        "CDB_USER_DATA": USER_DATA,
    }


def new_log_deployment_env():
    protocol = "docker://"
    return {
        "MERKELY_COMMAND": "log_deployment",
        "MERKELY_OWNER": OWNER,
        "MERKELY_PIPELINE": PIPELINE,
        "MERKELY_FINGERPRINT": f"{protocol}{IMAGE_NAME}",
        "MERKELY_API_TOKEN": API_TOKEN,
        "MERKELY_HOST": f"https://{DOMAIN}",

        "MERKELY_DESCRIPTION": DESCRIPTION,
        "MERKELY_ENVIRONMENT": ENVIRONMENT,
        "MERKELY_USER_DATA": USER_DATA,

        "BITBUCKET_WORKSPACE": BB_ORG,
        "BITBUCKET_REPO_SLUG": BB_REPO,
        "BITBUCKET_COMMIT": COMMIT,
        "BITBUCKET_BUILD_NUMBER": BUILD_NUMBER,
    }
