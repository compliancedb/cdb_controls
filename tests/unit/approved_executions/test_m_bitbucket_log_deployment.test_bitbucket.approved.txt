Getting SHA for docker image: acme/runner:4.56
Calculated digest: aacdaef69c676c2466571d3288880d559ccc2032b258fc5e73f99a103db462ee
Posting this payload:
{
    "artifact_sha256": "aacdaef69c676c2466571d3288880d559ccc2032b258fc5e73f99a103db462ee",
    "build_url": "https://bitbucket.org/acme/road-runner/addon/pipelines/home#!/results/1975",
    "description": "Deployment of acme/road-runner:4.67:aacdaef to docker hub",
    "environment": "dockerhub",
    "user_data": {
        "status": "deployed"
    }
}
To this url: https://app.compliancedb.com/api/v1/projects/acme/road-runner/deployments/
DRY RUN: Post not sent
